utils/densify.go

```diff
@@ -0,0 +1,137 @@
+package utils
+
+import (
+	"math"
+
+	geod "github.com/starboard-nz/go-geodesy"
+	"github.com/starboard-nz/orb"
+	"github.com/starboard-nz/units"
+)
+
+// DensifyMultiPolygon inserts points into the multipolygon using the given Model, until the maximum distance between
+// planar geometry and the given model is less than the tolerance.
+func DensifyMultiPolygon(mp orb.MultiPolygon, tolerance units.Distance, earthModel func(geod.LatLon, ...interface{}) geod.Model) orb.MultiPolygon {
+	var dmp orb.MultiPolygon
+
+	for _, polygon := range mp {
+		dmp = append(dmp, DensifyPolygon(polygon, tolerance, earthModel))
+	}
+
+	return dmp
+}
+
+// DensifyPolygon inserts points into the polygon using the given Model, until the maximum distance between
+// planar geometry and the given model is less than the tolerance.
+func DensifyPolygon(poly orb.Polygon, tolerance units.Distance, earthModel func(geod.LatLon, ...interface{}) geod.Model) orb.Polygon {
+	var dp orb.Polygon
+
+	for _, ring := range poly {
+		dp = append(dp, DensifyRing(ring, tolerance, earthModel))
+	}
+
+	return dp
+}
+
+// DensifyRing inserts points into the ring using the given Model, until the maximum distance between
+// planar geometry and the given model is less than the tolerance.
+func DensifyRing(ring orb.Ring, tolerance units.Distance, earthModel func(geod.LatLon, ...interface{}) geod.Model) orb.Ring {
+	if len(ring) < 2 {
+		return ring
+	}
+
+	lastPoint := ring[len(ring)-1]
+	closed := ring[0][0] == lastPoint[0] && ring[0][1] == lastPoint[1]
+
+	points := make([]orb.Point, 0, len(ring))
+	dr := orb.Ring(points)
+	dr = append(dr, ring[0])
+	for i := 1; i < len(ring); i++ {
+		ps := DensifySegment(ring[i-1], ring[i], tolerance, earthModel)
+		if len(ps) > 1 {
+			dr = append(dr, ps[1:]...)
+		} else {
+			// FIXME log error
+		}
+	}
+
+	if !closed {
+		ps := DensifySegment(lastPoint, ring[0], tolerance, earthModel)
+		if len(ps) > 1 {
+			dr = append(dr, ps[1:]...)
+		} else {
+			// FIXME log error
+		}
+	}
+
+	return dr
+}
+
+// DensifySegment inserts intermediate points into the segment p0-p1 using the given Model,
+// until the maximum distance between planar geometry and the given model is less than the tolerance.
+func DensifySegment(p0, p1 orb.Point, tolerance units.Distance, earthModel func(geod.LatLon, ...interface{}) geod.Model) []orb.Point {
+	if SegmentError(p0, p1, earthModel).Metre() <= tolerance.Metre() {
+		return []orb.Point{p0, p1}
+	}
```

Comment:
We should probably have another return value for `tolerance == 0` as it's currently possible to stack overflow.